pkg/driver/aci_driver.go

```diff
@@ -0,0 +1,770 @@
+		"ACI_DO_NOT_DELETE":        "Do not delete RG and ACI instance created - useful for debugging - only deletes RG if it was created by the driver",
+		"ACI_MSI_TYPE":             "If this is set to user or system the created ACI Container Group will be launched with MSI",
+		"ACI_SYSTEM_MSI_ROLE":      "The role to be asssigned to System MSI User - used if ACI_MSI_TYPE == system, if this is null or empty then the role defaults to contributor",
+		"ACI_SYSTEM_MSI_SCOPE":     "The scope to apply the role to System MSI User - will attempt to set scope to the  Resource Group that the ACI Instance is being created in if not set",
+		"ACI_USER_MSI_RESOURCE_ID": "The resource Id of the MSI User - required if ACI_MSI_TYPE == User",
+	}
+}
+
+// SetConfig sets ACI driver configuration
+func (d *ACIDriver) SetConfig(settings map[string]string) {
+	d.config = settings
+}
+
+// Run executes the ACI driver
+func (d *ACIDriver) Run(op *Operation) error {
+	return d.exec(op)
+}
+
+// Handles indicates that the ACI driver supports "docker" and "oci"
+func (d *ACIDriver) Handles(dt string) bool {
+	return dt == ImageTypeDocker || dt == ImageTypeOCI
+}
+
+func (d *ACIDriver) exec(op *Operation) (reterr error) {
+	d.out = op.Out
+	d.deleteACIResources = true
+	if len(d.config["ACI_DO_NOT_DELETE"]) > 0 && strings.ToLower(d.config["ACI_DO_NOT_DELETE"]) == "true" {
+		d.deleteACIResources = false
+	}
+
+	d.log("Delete Resources:", d.deleteACIResources)
+	d.inCloudShell = len(os.Getenv("ACC_CLOUD")) > 0
+	d.log("In Cloud Shell:", d.inCloudShell)
+	if d.Simulate {
+		return nil
+	}
+
+	err := d.loginToAzure()
+	if err != nil {
+		return fmt.Errorf("cannot Login To Azure: %v", err)
+	}
+
+	err = d.setAzureSubscriptionID()
+	if err != nil {
+		return fmt.Errorf("cannot set Azure subscription: %v", err)
+	}
+
+	err = d.createACIInstance(op)
+	if err != nil {
+		return fmt.Errorf("creating ACI instance failed: %v", err)
+	}
+
+	return nil
+}
+
+func (d *ACIDriver) loginToAzure() error {
+	clientID := d.config["AZURE_CLIENT_ID"]
+	clientSecret := d.config["AZURE_CLIENT_SECRET"]
+	tenantID := d.config["AZURE_TENANT_ID"]
+	applicationID := d.config["AZURE_APP_ID"]
+	d.log("clientID:", clientID)
+	d.log("clientSecret:", clientSecret)
+	d.log("tenantID:", tenantID)
+	d.log("applicationID:", applicationID)
+	// Attempt to login with Service Principal
+	if len(clientID) != 0 && len(clientSecret) != 0 && len(tenantID) != 0 {
+		d.log("Attempting to Login with Service Principal")
+		clientCredentailsConfig := auth.NewClientCredentialsConfig(clientID, clientSecret, tenantID)
+		authorizer, err := clientCredentailsConfig.Authorizer()
+		if err != nil {
+			return fmt.Errorf("Attempt to login to azure with Service Principal failed: %v", err)
+		}
+
+		d.authorizer = authorizer
+		return nil
+	}
+
+	// Attempt to login with Device Code
+	if len(applicationID) != 0 && len(tenantID) != 0 {
```

Comment:
This whole function sounds like it could be split into small pieces/sub-functions, with something like:
```go
switch{
case clientID != "" && clientSecret != "" && tenantID != "":
    return loginWithServicePrincipal(...)
case applicationID != "" && tenantID != "":
    return loginWithDeviceCode(...)
case d.inCloudShell:
    return loginWithCloudShell(...)
case checkForMSIEndpoint():
    return loginWithMSI(...)
default:
     return errors.New("Cannot login to Azure - no valid credentials provided")
}
```